CIP-0069

lehins opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins committed:
Add `PlutusArgs` data family. Remove PlutusScriptExamples

Theses tests are only checking the correctness of test plutus scripts
that we use. However, most of the future scripts that we will write will
depend on the TxInfo, which is not part of these tests.

lehins committed:
Update cabal bounds for plutus-ledger-api

lehins committed:
Regenerate golden tests for script translation

lehins committed:
Remove unusable and redundant plutus sample scripts

lehins committed:
Rename sample plutus scripts:

* `alwaysSucceeds2` -> `alwaysSucceedsNoDatum`
* `alwaysSucceeds3` -> `alwaysSucceedsWithDatum`
* `alwaysFails2` -> `alwaysFailsNoDatum`
* `alwaysFails3` -> `alwaysFailsWithDatum`
* `guessTheNumber3` -> `redeemerSameAsDatum`
* `evendata3` -> `evenDatum`
* `evenRedeemer2` -> `evenRedeemerNoDatum`
* `evenRedeemer3` -> `evenRedeemerWithDatum`

* Regenerate plutus sample scripts with new names
* Apply the plutus sample script renaming to the test suites

lehins committed:
plutus-preprocessor improvements:

* Remove dead code

* Rearrange ...

* Fix #4230 implement hex printing

lehins committed:
Add `eraLanguages` and improve UTXOW imp tests

lehins committed:
Add `ExtraRedeemers/RedeemerPointerPointsToNothing` imp test for Babbage

lehins committed:
Change zero valued testing cost models

Plutus has switched to variable number of parameters approach and
increates the number of parameters for PlutusV2. Moreover testing cost
models on the plutus side have not cought up yet.

In any case, until we are in conway we can't vary number of parameters
on the ledger side for PlutusV1/PlutusV2, thereforethus the switch to a
hardcoded number of cost model parameters.

Temporarely make test cost model contain all zeros.

lehins committed:
Reduce number of CostModel params for PlutusV3 to 231

lehins committed:
Remove changelogs for packages that are not subject to release

lehins committed:
Step 1/2: Duplicate and move modules

lehins committed:
Step 2/2. Finish module restructure

lehins committed:
Fix and improve PlutusV1/V2/V3 test scripts:

* Add haddock

* Add ScriptHash to every plutus test example

* Make test script much stricter in the way they fail

lehins committed:
Fix a bunch of failures due to invalid script being used

lehins committed:
Enhance Alonzo's UTXOS imp test to check all plutus languages

lehins committed:
Add `Uniform` and `UniformRange` instances for `Coin`/`CompactCoin`

lehins committed:
Add a test for a Spending PlutusV3 script that requires no datum

lehins committed:
Allow PlutusV3 spending without Datums

Add a new version of deprecate old `lookupPlutusScript`

lehins committed:
Add ability to test script failure

lehins committed:
Fix datum addition fixup to the TxOut

lehins committed:
Update HaskellNix

aniketd committed:
Write tests for purposeIsWellformed for V1,2,3

teodanciu reviewed (commented):
Suggestion for cardano-ledger-alonzo  CHANGELOG: 
 * Add `LedgerTxInfo` to `Cardano.Ledger.Alonzo.Plutus.Context`
 * `EraPlutusTxInfo` class:
 	* replace params in `toPlutusTxInfo` signature with `LedgerTxInfo`
 	* replace `toPlutusScriptContext` with `toPlutusArgs` 
 * `EraPlutusContext` class:
 	* remove `mkPlutusScriptContext`, `mkPlutusLanguageContext`
 	* add `mkPlutusWithContext` 
 * add `toPlutusWithContext` to `Cardano.Ledger.Alonzo.Plutus.Context` 
 * Move `PlutusScriptContext` type family to core
 * Add `ContextError` inhabitant to `TransactionScriptFailure`
 * Change return type of `evalTxExUnits` and `evalTxExUnitsWithLogs` to not be `Either`s 
 * Add `lookupPlutusScript` and `eraLanguages` to Alonzo `Scripts`
 * Deprecate `lookupPlutusScript` in Alonzo `Evaluate`

teodanciu reviewed (commented):
Suggestions for `cardano-ledger-api` CHANGELOG: 
 * `TranslationInstance`: replace PParams field with ProtVer

teodanciu reviewed (commented):
Suggestion for cardano-ledger-conway CHANGELOG:

* change return type of `trySubmitTxto`  to include the fixedup tx

lehins committed:
Switch to BuiltinUnit for PlutusV3 scripts

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -350,9 +355,9 @@ evalTxExUnits ::
   --  sufficient execution budget.
   --  Otherwise, we return a 'TranslationError' manifesting from failed attempts
   --  to construct a valid execution context for the given transaction.
```
```suggestion
  --  sufficient execution budget.
```

teodanciu replied:
Same below, for `evalTxExUnitsWithLogs` here:  https://github.com/IntersectMBO/cardano-ledger/pull/4374/files#diff-d42005a7dd7c31638237a88fcf0ffe4feb1cca752f80b3d8901d8626e6d10f45L379

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -350,9 +355,9 @@ evalTxExUnits ::
+  Map.map (fmap snd) $ evalTxExUnitsWithLogs pp tx utxo epochInfo systemStart
 
 -- | Evaluate the execution budgets needed for all the redeemers in
 --  a given transaction. If a redeemer is invalid, a failure is returned instead.
```
```suggestion
-- | Evaluate the execution budgets needed for all the redeemers in
--  a given transaction.
```

teodanciu replied:
Same comment about the comment above for `evalTxExUnits` here: https://github.com/IntersectMBO/cardano-ledger/pull/4374/files#diff-d42005a7dd7c31638237a88fcf0ffe4feb1cca752f80b3d8901d8626e6d10f45R334

teodanciu commented on libs/cardano-ledger-core/CHANGELOG.md:
```diff
@@ -6,6 +6,19 @@
 * Remove unnecessary instances: `Inject a ()` and `Inject Void a`
 * Add `integralToBounded` to `BaseTypes`
```
```suggestion
* Add `integralToBounded` to `BaseTypes`
* Add `PlutusScriptContext` type family moved from Alonzo
* Add `LegacyPlutusArgs` data
* Add `PlutusArgs` data family to `PlutusLanguage` and instances
* Change signature of `evaluatePlutusRunnable` and `evaluatePlutusRunnableBudget` to replace data list with `PlutusArgs`

```

lehins committed:
Add two versions of inputsOutputsAreNotEmpty

lehins committed:
Make the demon of old shelley generators happy

Decrease `maxMinFeeA` in `Constants`

used to generate PParams for old chain trace generators, so that we don't discard traces when there is not enough money to pay the fee

teodanciu commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Utxo.hs:
```diff
@@ -387,7 +387,7 @@ genNextDelta
         draftSize =
           sum
-            [ 1100 :: Integer -- safety net in case the coin or a list prefix rolls over into a
+            [ 11000 :: Integer -- safety net in case the coin or a list prefix rolls over into a
```
I have experimented with decreasing generated minFeeA and minFeeB - if we dramatically decrease it here in [Constants](https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Constants.hs#L89) , as  I did [in this commit](https://github.com/IntersectMBO/cardano-ledger/commit/01f01f68be632f9dac8e186d9520b4ab42006a46), tests pass reliably. 
We could perhaps do a combination of decreasing (maybe less drastically) minFee and increasing whatever this draftSize is.  I have noticed that with your change to 11000 , we can have maxMinFeeA  to be 100 and the number of discarded tests is still small.

teodanciu replied:
There is another failing test because of discarded traces in shelley: https://github.com/IntersectMBO/cardano-ledger/actions/runs/9505453394/job/26203097744#step:12:62

Lowering  maxMinFeeA and maxMinFeeB as above also leads to running these tests reliably (usually under 5 discarded traces, compared to about 200 which is with the current values).

lehins replied:
Thank you! :slightly_smiling_face:

teodanciu commented on libs/plutus-preprocessor/app/Main.hs:
```diff
@@ -0,0 +1,17 @@
+-- package to build the system.  If the plutus package changes, we will have to regenerate
+-- the Examples.hs file.
+-- To regenerate Examples.hs, on a machine that can depend upon plutus=plugin,
+-- run 'cabal run plutus-preprocessor'
```
```suggestion
-- run 'cabal run plutus-preprocessor:plutus-preprocessor' using ghc = 9.6.x
```

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -382,76 +387,55 @@ evalTxExUnitsWithLogs ::
+evalTxExUnitsWithLogs pp tx utxo epochInfo systemStart = Map.mapWithKey findAndCount rdmrs
   where
+    usePointerAsKey (plutusPurpose, _) = hoistPlutusPurpose toAsIx plutusPurpose
+    ptrToPlutusScript = fromElems usePointerAsKey scriptsNeeded
```
I find the pointer/ptr terminology quite confusing. What do you think about:
`keyedByPurpose` instead of `usePointerAsKey` and `purposeToScriptHash` instead  of `ptrToPlutusScript` ?
```suggestion
    keyedByPurpose (plutusPurpose, _) = hoistPlutusPurpose toAsIx plutusPurpose
    purposeToScriptHash = fromElems keyedByPurpose scriptsNeeded
```
Of course there might be a reason behind this that I'm not seeing, so please ignore if that's the case

teodanciu commented:
It seems that default cost models for testing  are not yet available via plutus-ledger-api, so we will merge this with zero cost models for tests and update them when they will become available. Added [an issue](https://github.com/IntersectMBO/cardano-ledger/issues/4404) to record this

teodanciu reviewed (approved):
The changes seem as if they are "obvious in hindsight" - i think it's a good sign that it's improving upon what we have, awesome work!!  :star_struck: 
Looks good to me, just some small remarks.

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -382,76 +387,55 @@ evalTxExUnitsWithLogs ::
+                  , sh
+                  )
+              )
               ptrToPlutusScript
```
```suggestion
              purposeToScriptHash
```
if you took my suggestion above

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -382,76 +387,55 @@ evalTxExUnitsWithLogs ::
+    findAndCount pointer (redeemerData, exUnits) = do
+      (plutusPurpose, plutusScriptHash) <-
         note (RedeemerPointsToUnknownScriptHash pointer) $
           Map.lookup pointer ptrToPlutusScript
```
```suggestion
          Map.lookup pointer purposeToScriptHash
```
if you took my suggestion above

teodanciu committed:
Simplify `evalPlutusScripts` to remove unused tx parameter
